x-pack/qa/kerberos-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/CustomHttpClientConfigCallbackHandler.java

```diff
@@ -0,0 +1,338 @@
+            }
+            final Path path = PathUtils.get(keystorePath);
+            if (!Files.exists(path)) {
+                throw new IllegalStateException(ESRestTestCase.TRUSTSTORE_PATH + " is set but points to a non-existing file");
+            }
+            try (InputStream is = Files.newInputStream(path)) {
+                final KeyStore keyStore = KeyStore.getInstance("jks");
+                keyStore.load(is, keystorePass.toCharArray());
+                final SSLContext sslcontext = SSLContexts.custom().loadTrustMaterial(keyStore, null).build();
+                final SSLIOSessionStrategy sessionStrategy = new SSLIOSessionStrategy(sslcontext);
+                httpClientBuilder.setSSLStrategy(sessionStrategy);
+            } catch (KeyStoreException | NoSuchAlgorithmException | KeyManagementException | CertificateException e) {
+                throw new RuntimeException("Error setting up ssl", e);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private void setupSpnegoAuthSchemeSupport(HttpAsyncClientBuilder httpClientBuilder) {
+        final Lookup<AuthSchemeProvider> authSchemeRegistry = RegistryBuilder.<AuthSchemeProvider>create()
+                .register(AuthSchemes.SPNEGO, new SPNegoSchemeFactory()).build();
+
+        final GSSManager gssManager = GSSManager.getInstance();
+        try {
+            final GSSName gssUserPrincipalName = gssManager.createName(userPrincipalName, GSSName.NT_USER_NAME);
+            login();
+            final GSSCredential credential = Subject.doAs(loginContext.getSubject(),
+                    (PrivilegedExceptionAction<GSSCredential>) () -> gssManager.createCredential(gssUserPrincipalName,
+                            GSSCredential.DEFAULT_LIFETIME, SPNEGO_OID, GSSCredential.INITIATE_ONLY));
+
+            final KerberosCredentialsProvider credentialsProvider = new KerberosCredentialsProvider();
+            credentialsProvider.setCredentials(
+                    new AuthScope(AuthScope.ANY_HOST, AuthScope.ANY_PORT, AuthScope.ANY_REALM, AuthSchemes.SPNEGO),
+                    new KerberosCredentials(credential));
+            httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+        } catch (GSSException e) {
+            throw new RuntimeException(e);
+        } catch (LoginException e) {
+            throw new RuntimeException(e);
+        } catch (PrivilegedActionException e) {
+            throw new RuntimeException(e);
+        }
+        httpClientBuilder.setDefaultAuthSchemeRegistry(authSchemeRegistry);
+    }
+
+    /**
+     * If logged in {@link LoginContext} is not available, it attempts login and
+     * returns {@link LoginContext}
+     *
+     * @return {@link LoginContext}
+     * @throws LoginException
+     */
+    public LoginContext login() throws LoginException {
```

Comment:
can you make this synchronized? There is no concurrent caller but I don't like having a public API like this since someone could call it concurrently